jlm/util/GraphWriter.hpp

```diff
@@ -0,0 +1,843 @@
+   * Depending on output format, this function may also recurse and print sub graphs.
+   */
+  void
+  Output(std::ostream & out, GraphOutputFormat format, size_t indent);
+
+  /**
+   * Outputs the node in ASCII format to the ostream \p out, indented by \p indent levels.
+   * In this format, attributes are ignored, and edges are only included implicitly,
+   * by listing the origins of all edges pointing into this node.
+   */
+  virtual void
+  OutputASCII(std::ostream & out, size_t indent);
+
+  /**
+   * Outputs the node in Dot format to the ostream \p out, indented by \p indent levels.
+   * This format includes all attributes. Edges are output
+   */
+  virtual void
+  OutputDot(std::ostream & out, size_t indent);
+
+  /**
+   * Prints all sub graphs of the node, to the given ostream \p out, in the given \p format.
+   * * All lines printed by this function are indented by at least \p indent levels.
+   * This function is recursive, as sub graphs may have nodes with sub graphs of their own.
+   */
+  virtual void
+  OutputSubgraphs(std::ostream & out, GraphOutputFormat format, size_t indent);
+
+private:
+  Graph & Graph_;
+};
+
+// InOut nodes represent operations with clearly defined inputs and outputs, with ports for each
+class InOutNode;
+class InputPort;
+class OutputPort;
+
+/**
+ * The input port of an InOutNode
+ */
+class InputPort : public Port
+{
+  friend InOutNode;
+
+  InputPort(InOutNode & node);
+
+public:
+  ~InputPort() override = default;
+
+  const char *
+  GetIdPrefix() const override;
+
+  Node &
+  GetNode() override;
+
+  bool
+  CanBeEdgeTail() const override;
+
+  void
+  SetFillColor(std::string color) override;
+
+  void
+  OutputDotPortId(std::ostream & out) override;
+
+private:
+  InOutNode & Node_;
+};
+
+class OutputPort : public Port
+{
+  friend InOutNode;
+
+  OutputPort(InOutNode & node);
+
+public:
+  ~OutputPort() override = default;
+
+  const char *
+  GetIdPrefix() const override;
+
+  Node &
+  GetNode() override;
+
+  bool
+  CanBeEdgeHead() const override;
+
+  void
+  SetFillColor(std::string color) override;
+
+  void
+  OutputDotPortId(std::ostream & out) override;
+
+private:
+  InOutNode & Node_;
+};
+
+class InOutNode : public Node
+{
+  friend Graph;
+
+  InOutNode(Graph & graph, size_t inputPorts, size_t outputPorts);
+
+public:
+  ~InOutNode() override = default;
+
+  InputPort &
+  CreateInputPort();
+
+  size_t
+  NumInputPorts();
+
+  InputPort &
+  GetInputPort(size_t index);
+
+  OutputPort &
+  CreateOutputPort();
+
+  size_t
+  NumOutputPorts();
+
+  OutputPort &
+  GetOutputPort(size_t index);
+
+  Graph &
+  CreateSubgraph();
+
+  size_t
+  NumSubgraphs();
+
+  Graph &
+  GetSubgraph(size_t index);
+
+  /**
+   * Regular attributes are placed on the dot node itself.
+   * Use this method to set attributes on the HTML table printed when outputting InOutNodes to dot.
+   * The attributes will affect the outmost table with a visible border.
```

Comment:
What is the outmost table?

One needs to understand the layout in order to use this method.